### PR TITLE
Adds event term to randomize gains of explicit actuators

### PR DESCRIPTION
**Pull request by MoreTore**
# Description
resolves: https://github.com/isaac-sim/IsaacLab/issues/1000
resolves: https://github.com/isaac-sim/IsaacLab/issues/1011
depends: https://github.com/isaac-sim/IsaacLab/issues/1006

## Type of change
- New feature (non-breaking change which adds functionality)
- This change may require a documentation update

## Checklist

- [x] have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [x] My changes generate no new warnings
- [x] I have added my name to the `CONTRIBUTORS.md` or my name already exists there
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file

### CODE REVIEW

**Comment by MoreTore** [comment]
Can I have some guidance on how the test should be completed

**Comment by MoreTore** [comment]
in articulation.py we could set the default_joint_stiffness and default_joint_damping properly , instead of settings all zeros for explicit actuators. this would simplify this PRs code but its out of scope for this small change.

but for example in `source/extensions/omni.isaac.lab/omni/isaac/lab/assets/articulation/articulation.py`:
````python
for actuator_name, actuator_cfg in self.cfg.actuators.items():
  if isinstance(actuator, ImplicitActuator):
     self.write_joint_xxx
     ...
  else:
     self.write_joint_xxx
     ...
     self._data.default_joint_stiffness[slice(None), actuator.joint_indices] = (
        actuator._parse_joint_parameter(actuator.cfg.stiffness, actuator.stiffness)
     )
     self._data.default_joint_damping[slice(None), actuator.joint_indices] = (
        actuator._parse_joint_parameter(actuator.cfg.damping, actuator.damping)
     )
self._data.default_joint_stiffness[implicit indexes only] = self.root_physx_view.get_dof_stiffnesses().to(device=self.device).clone()[implicit indexes only]
self._data.default_joint_damping[implicit indexes only] = self.root_physx_view.get_dof_damping().to(device=self.device).clone()[implicit indexes only]
````

**Comment by MoreTore** [comment]
Added my proposed changes to articulations.py to this PR.

**Comment by MoreTore** [comment]
Damn tensor indexing and slicing is hard
```
../aten/src/ATen/native/cuda/IndexKernel.cu:92: operator(): block: [0,0,0], thread: [0,0,0] Assertion `-sizes[i] <= index && index < sizes[i] && "index out of bounds"` failed.
../aten/src/ATen/native/cuda/IndexKernel.cu:92: operator(): block: [0,0,0], thread: [1,0,0] Assertion `-sizes[i] <= index && index < sizes[i] && "index out of bounds"` failed.
../aten/src/ATen/native/cuda/IndexKernel.cu:92: operator(): block: [0,0,0], thread: [2,0,0] Assertion `-sizes[i] <= index && index < sizes[i] && "index out of bounds"` failed.
../aten/src/ATen/native/cuda/IndexKernel.cu:92: operator(): block: [0,0,0], thread: [3,0,0] Assertion `-sizes[i] <= index && index < sizes[i] && "index out of bounds"` failed.
../aten/src/ATen/native/cuda/IndexKernel.cu:92: operator(): block: [0,0,0], thread: [4,0,0] Assertion `-sizes[i] <= index && index < sizes[i] && "index out of bounds"` failed.
../aten/src/ATen/native/cuda/IndexKernel.cu:92: operator(): block: [0,0,0], thread: [5,0,0] Assertion `-sizes[i] <= index && index < sizes[i] && "index out of bounds"` failed.
../aten/src/ATen/native/cuda/IndexKernel.cu:92: operator(): block: [0,0,0], thread: [6,0,0] Assertion `-sizes[i] <= index && index < sizes[i] && "index out of bounds"` failed.
../aten/src/ATen/native/cuda/IndexKernel.cu:92: operator(): block: [0,0,0], thread: [7,0,0] Assertion `-sizes[i] <= index && index < sizes[i] && "index out of bounds"` failed.
```

**Comment by MoreTore** [comment]
still having bugs

File: source/extensions/omni.isaac.lab/omni/isaac/lab/actuators/actuator_base.py [hunk 1]
```diff
@@ -208,27 +208,32 @@ def _parse_joint_parameter(
...
+                param = torch.full((self._num_envs, self.num_joints), float(default_value), device=self._device)
             elif isinstance(default_value, torch.Tensor):
-                # if tensor, then use the same tensor for all joints
-                param[:] = default_value.float()
+                # Use the default tensor directly if shapes match, or copy values otherwise
+                if default_value.shape == (self._num_envs, self.num_joints):
+                    param = default_value.float()
+                else:
+                    param = torch.zeros(self._num_envs, self.num_joints, device=self._device)
+                    param[:] = default_value.float()
```
Comment (Dhoeller19) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/actuators/actuator_base.py]: Maybe it is worth throwing an error here?

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py [hunk 2]
```diff
@@ -264,53 +264,56 @@ def randomize_actuator_gains(
...
+    def randomize(data: torch.Tensor, params: tuple[float, float]) -> torch.Tensor:
+        return _randomize_prop_by_op(
+            data, params, env_ids, target_joint_indices, operation=operation, distribution=distribution
+        )
+
+    # Loop through actuators and randomize gains
+    for actuator in asset.actuators.values():
+        # Resolve joint indices for each actuator
+        if isinstance(actuator.joint_indices, slice):  # All joints in this actuator
+            target_joint_indices = torch.arange(asset.num_joints, dtype=torch.int, device=asset.device)
```
Comment (Dhoeller19) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py]: Shouldn't there also be an intersection here?

**Comment by Dhoeller19** [comment]
This is great stuff, thanks @MoreTore!

**Comment by MoreTore** [comment]
@Dhoeller19 Thanks for the review! I have a newer version that I have been using. I will incorporate the suggestions you noted.

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py [hunk 3]
```diff
@@ -263,55 +263,51 @@ def randomize_actuator_gains(
...
+        all_envs = torch.arange(env.scene.num_envs, device=asset.device)
+        # Randomize stiffness
+        if stiffness_distribution_params is not None:
+            actuator.stiffness = asset.data.default_joint_stiffness.to(asset.device).clone()
+            sim_stiffness = randomize(
+                actuator.stiffness, stiffness_distribution_params
+            )[env_ids][:, target_joint_indices]
+            actuator.stiffness = actuator.stiffness[all_envs[:, None], actuator.joint_indices]
+            if isinstance(actuator, ImplicitActuator):
+                asset.write_joint_stiffness_to_sim(sim_stiffness, joint_ids=target_joint_indices, env_ids=env_ids)
```
Comment (Dhoeller19) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py]:
```suggestion
            randomize(actuator.stiffness, stiffness_distribution_params)
            actuator.stiffness = actuator.stiffness[all_envs[:, None], actuator.joint_indices]
            if isinstance(actuator, ImplicitActuator):
                asset.write_joint_stiffness_to_sim(actuator.stiffness[env_ids][:, target_joint_indices], joint_ids=target_joint_indices, env_ids=env_ids)
```

Comment (Dhoeller19) [reply source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py]: Is this simpler to understand maybe?

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py [hunk 4]
```diff
@@ -263,55 +263,51 @@ def randomize_actuator_gains(
...
+        # Randomize damping
+        if damping_distribution_params is not None:
+            actuator.damping = asset.data.default_joint_damping.to(asset.device).clone()
+            sim_damping = randomize(actuator.damping, damping_distribution_params)[env_ids][:, target_joint_indices]
+            actuator.damping = actuator.damping[all_envs[:, None], actuator.joint_indices]
+            if isinstance(actuator, ImplicitActuator):
+                asset.write_joint_damping_to_sim(sim_damping, joint_ids=target_joint_indices, env_ids=env_ids)
```
Comment (Dhoeller19) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py]:
```suggestion
            randomize(actuator.damping, damping_distribution_params)
            actuator.damping = actuator.damping[all_envs[:, None], actuator.joint_indices]
            if isinstance(actuator, ImplicitActuator):
                asset.write_joint_damping_to_sim(actuator.damping[env_ids][:, target_joint_indices], joint_ids=target_joint_indices, env_ids=env_ids)
```

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py [hunk 5]
```diff
@@ -263,55 +263,51 @@ def randomize_actuator_gains(
...
+            continue
+        
+        all_envs = torch.arange(env.scene.num_envs, device=asset.device)
+        # Randomize stiffness
+        if stiffness_distribution_params is not None:
+            actuator.stiffness = asset.data.default_joint_stiffness.to(asset.device).clone()
+            sim_stiffness = randomize(
+                actuator.stiffness, stiffness_distribution_params
+            )[env_ids][:, target_joint_indices]
+            actuator.stiffness = actuator.stiffness[all_envs[:, None], actuator.joint_indices]
```
Comment (Dhoeller19) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py:301]: Is this reshaping really needed now?

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py [hunk 6]
```diff
@@ -263,55 +263,51 @@ def randomize_actuator_gains(
...
+            if isinstance(actuator, ImplicitActuator):
+                asset.write_joint_stiffness_to_sim(sim_stiffness, joint_ids=target_joint_indices, env_ids=env_ids)
+        # Randomize damping
+        if damping_distribution_params is not None:
+            actuator.damping = asset.data.default_joint_damping.to(asset.device).clone()
+            sim_damping = randomize(actuator.damping, damping_distribution_params)[env_ids][:, target_joint_indices]
+            actuator.damping = actuator.damping[all_envs[:, None], actuator.joint_indices]
```
Comment (Dhoeller19) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py:308]: Is this reshaping really needed now?

File: source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py [hunk 7]
```diff
@@ -263,55 +263,49 @@ def randomize_actuator_gains(
...
+            actuator.stiffness = actuator.stiffness[all_envs[:, None], actuator.joint_indices]
+            if isinstance(actuator, ImplicitActuator):
+                asset.write_joint_stiffness_to_sim(actuator.stiffness[env_ids][:, target_joint_indices], joint_ids=target_joint_indices, env_ids=env_ids)
+        # Randomize damping
+        if damping_distribution_params is not None:
+            actuator.damping = asset.data.default_joint_damping.to(asset.device).clone()
+            randomize(actuator.damping, damping_distribution_params)
+            actuator.damping = actuator.damping[all_envs[:, None], actuator.joint_indices]
+            if isinstance(actuator, ImplicitActuator):
+                asset.write_joint_damping_to_sim(actuator.damping[env_ids][:, target_joint_indices], joint_ids=target_joint_indices, env_ids=env_ids)
```
Comment (Dhoeller19) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py]:
```suggestion
        # Randomize stiffness
        if stiffness_distribution_params is not None:
            stiffness = asset.data.default_joint_stiffness.to(asset.device).clone()
            randomize(stiffness, stiffness_distribution_params)
            actuator.stiffness = stiffness[all_envs[:, None], actuator.joint_indices]
            if isinstance(actuator, ImplicitActuator):
                asset.write_joint_stiffness_to_sim(actuator.stiffness[env_ids[:, None], target_joint_indices], joint_ids=target_joint_indices, env_ids=env_ids)
        # Randomize damping
        if damping_distribution_params is not None:
            damping = asset.data.default_joint_damping.to(asset.device).clone()
            randomize(damping, damping_distribution_params)
            actuator.damping = damping[all_envs[:, None], actuator.joint_indices]
            if isinstance(actuator, ImplicitActuator):
                asset.write_joint_damping_to_sim(actuator.damping[env_ids[:, None], target_joint_indices], joint_ids=target_joint_indices, env_ids=env_ids)
```

Comment (Dhoeller19) [reply source/extensions/omni.isaac.lab/omni/isaac/lab/envs/mdp/events.py]: @MoreTore I just noticed before, I would modify the `actuator.stiffness` and `actuator.damping` values for all the envs which is not what we want. This suggestion should fix it, could you quickly double check?

**Comment by Dhoeller19** [comment]
@MoreTore sorry that it took so long. We can merge as soon as you check the last conversation.

**Comment by MoreTore** [comment]
> @MoreTore sorry that it took so long. We can merge as soon as you check the last conversation and formatting is run with `./isaaclab.sh -f`

Yea the issue now will be making sure the merge doesn't break the other changes I made :p

**Review by Dhoeller19 — APPROVED** [review]

**Pull request merged**